### ISSUE TITLE
DataFile behaves differently inside with statement

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#1242]
I noticed while looking at #1241 that DataFile has unexpected (to me) behaviour inside a with statement. https://github.com/boutproject/BOUT-dev/blob/5dcf4041d6c43a00d29d69c1b08fca2c36d78bc4/tools/pylib/boututils/datafile.py#L166-L167
because `__enter__` returns `self.impl.__enter__()`, in a statement like
```
with DataFile('BOUT.dmp.0.nc') as f:
    # do stuff here
```
`f` is a `DataFile_netCDF` object, not a `DataFile` object. In practice this doesn't make much difference because the interfaces are very similar, but `DataFile.read` does some work
https://github.com/boutproject/BOUT-dev/blob/5dcf4041d6c43a00d29d69c1b08fca2c36d78bc4/tools/pylib/boututils/datafile.py#L197-L201
(although I think both netCDF and hdf5 can handle converting lists and tuples to slices) and `bout_type(self, varname)` which returns `self.attributes(varname)["bout_type"]` would be missing.

In summary, I'd expect `DataFile.__enter__` to return `self` so that `f` in the `with`-statement is a `DataFile` object not one of the implementations. Is this a bug?

**Comment by ZedThree** [comment]
I think with our current implementation, `DataFile.__enter__` should call `self.impl.__enter__()` but return `self`. Any public methods that the implementations define should also be defined in `DataFile`, which can just call `self.impl.foo()`

**Comment by dschwoerer** [comment]
Fixed by #1243
